File: SalesManagementSystem/fix_encoding.js
const fs = require('fs');
const path = require('path');

function getFiles(dir, files_) {
    files_ = files_ || [];
    const files = fs.readdirSync(dir);
    for (let i in files) {
        const name = dir + '/' + files[i];
        if (fs.statSync(name).isDirectory()) {
            if (!name.includes('node_modules') && !name.includes('.git') && !name.includes('obj') && !name.includes('bin')) {
                getFiles(name, files_);
            }
        } else {
            if (name.endsWith('.cs') || name.endsWith('.cshtml')) {
                files_.push(name);
            }
        }
    }
    return files_;
}

const allFiles = getFiles('.');
let totalFixed = 0;

function fixString(s) {
    if (s.includes('Ã') || s.includes('Ä') || s.includes('á') || s.includes('Æ')) {
        try {
            const decoded = Buffer.from(s, 'latin1').toString('utf8');
            if (!decoded.includes('\ufffd')) {
                return decoded;
            }
        } catch (e) {}
    }
    return null;
}

for (const file of allFiles) {
    let originalText = fs.readFileSync(file, 'utf8');
    let fixedText = originalText;

    // Fix strings in quotes
    fixedText = fixedText.replace(/"([^"\\]*(?:\\.[^"\\]*)*)"/g, (match, p1) => {
        const fixed = fixString(p1);
        if (fixed) return '"' + fixed + '"';
        return match;
    });

    if (file.endsWith('.cshtml')) {
        // Fix text between HTML tags
        fixedText = fixedText.replace(/>([^<]+)</g, (match, p1) => {
            const fixed = fixString(p1);
            if (fixed) return '>' + fixed + '<';
            return match;
        });
    }

    if (originalText !== fixedText) {
        fs.writeFileSync(file, fixedText, 'utf8');
        console.log('Fixed ' + file);
        totalFixed++;
    }
}
console.log('Total fixed: ' + totalFixed);

File: SalesManagementSystem/fix_encoding.py
import os
import re

count = 0
for root, dirs, files in os.walk('.'):
    dirs[:] = [d for d in dirs if d not in ('bin', 'obj', '.git', 'node_modules')]
    for file in files:
        if file.endswith('.cs') or file.endswith('.cshtml'):
            path = os.path.join(root, file)
            with open(path, 'r', encoding='utf-8', errors='ignore') as f:
                content = f.read()
            
            if 'Ä' in content or 'Ã' in content or 'á' in content or 'Æ' in content:
                def repl(m):
                    s = m.group(1)
                    if 'Ä' in s or 'Ã' in s or 'á' in s or 'Æ' in s:
                        try:
                            decoded = s.encode('cp1252').decode('utf-8')
                            if '\ufffd' not in decoded:
                                return '"' + decoded + '"'
                        except:
                            pass
                    return m.group(0)
                
                new_content = re.sub(r'"([^"\\]*(?:\\.[^"\\]*)*)"', repl, content)
                
                if file.endswith('.cshtml'):
                    def repl_html(m):
                        s = m.group(1)
                        if 'Ä' in s or 'Ã' in s or 'á' in s or 'Æ' in s:
                            try:
                                decoded = s.encode('cp1252').decode('utf-8')
                                if '\ufffd' not in decoded:
                                    return '>' + decoded + '<'
                            except:
                                pass
                        return m.group(0)
                    new_content = re.sub(r'>([^<]+)<', repl_html, new_content)
                
                if new_content != content:
                    with open(path, 'w', encoding='utf-8') as f:
                        f.write(new_content)
                    print('Fixed', path)
                    count += 1
print('Total fixed:', count)
